nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchHttp.java

```diff
@@ -0,0 +1,330 @@
+                    sb.append(index);
+                    sb.append("\", \"_type\": \"");
+                    sb.append(docType);
+                    sb.append("\"");
+                    if (!StringUtils.isEmpty(id)) {
+                        sb.append(", \"_id\": \"");
+                        sb.append(id);
+                        sb.append("\"");
+                    }
+                    sb.append("}}\n");
+                    sb.append(json);
+                    sb.append("\n");
+                } else if (indexOp.equalsIgnoreCase("upsert") || indexOp.equalsIgnoreCase("update")) {
+                    sb.append("{\"update\": { \"_index\": \"");
+                    sb.append(index);
+                    sb.append("\", \"_type\": \"");
+                    sb.append(docType);
+                    sb.append("\", \"_id\": \"");
+                    sb.append(id);
+                    sb.append("\" }\n");
+                    sb.append("{\"doc\": ");
+                    sb.append(json);
+                    sb.append(", \"doc_as_upsert\": ");
+                    sb.append(indexOp.equalsIgnoreCase("upsert"));
+                    sb.append(" }\n");
+                } else if (indexOp.equalsIgnoreCase("delete")) {
+                    sb.append("{\"delete\": { \"_index\": \"");
+                    sb.append(index);
+                    sb.append("\", \"_type\": \"");
+                    sb.append(docType);
+                    sb.append("\", \"_id\": \"");
+                    sb.append(id);
+                    sb.append("\" }\n");
+                }
+            }
+            if (!flowFilesToTransfer.isEmpty()) {
+                RequestBody requestBody = RequestBody.create(MediaType.parse("application/json"), sb.toString());
+                final Response getResponse = sendRequestToElasticsearch(okHttpClient, url, "PUT", requestBody);
+                final int statusCode = getResponse.code();
+
+                if (isSuccess(statusCode)) {
+                    ResponseBody responseBody = getResponse.body();
+                    final byte[] bodyBytes = responseBody.bytes();
+                    JsonNode responseJson = parseJsonResponse(new ByteArrayInputStream(bodyBytes));
+                    boolean errors = responseJson.get("errors").asBoolean(false);
+                    if (errors) {
+                        ArrayNode itemNodeArray = (ArrayNode) responseJson.get("items");
+                        if (itemNodeArray.size() > 0) {
+                            // All items are returned whether they succeeded or failed, so iterate through the item array
+                            // at the same time as the flow file list, moving each to success or failure accordingly
+                            for (int i = 0; i < itemNodeArray.size(); i++) {
+                                JsonNode itemNode = itemNodeArray.get(i);
+                                FlowFile flowFile = flowFilesToTransfer.remove(i);
```

Comment:
Appears to be but the doc doesn't explicitly guarantee it: "The response to a bulk action is a large JSON structure with the individual results of each action that was performed. The failure of a single action does not affect the remaining actions." However if the results were out of order, it would be extremely difficult for any client to realign the operations with the results.